### DISCUSSION TITLE
Double free or corruption (out) error

### DISCUSSION THREAD

**Discussion by enahon** [borglab/gtsam#1450] [Q&A] [answered]
## Description

I am linking GTSAM to an executable which also uses PCL and Eigen. When I launch the executable, the program immediately crashes with the following error: "Double free or corruption (out)".

## Environment

I am running this on Ubuntu 22.04 and using GCC 11.3.0. I have built GTSAM v4.1.1 and PCL 1.13.0 from source. I am also using Eigen 3.4.0.

I have built GTSAM using the following options:
        -DGTSAM_USE_SYSTEM_EIGEN=ON
        -DGTSAM_BUILD_WITH_MARCH_NATIVE=OFF
I tried enabling DGTSAM_BUILD_WITH_MARCH_NATIVE but the same error occurs. Upgrading GTSAM to the latest 4.2a9 pre-release didn't help either.
I have also tried adding the following option when building GTSAM: 
        -DCMAKE_CXX_FLAGS:STRING=-std=c++17

I am building PCL with default options.

Any ideas will be much appreciated!
Thanks.

## Additional information

I have tried building GTSAM in Debug mode and I am able to see that it's actually crashing during GTSAM initialization inside Eigen's DenseStorage.h, line 147, when calling EIGEN_MAKE_UNALIGNED_ARRAY_ASSERT(63). 

Call stack:

- libc.so.6!__GI___assert_fail(const char * assertion, const char * file, unsigned int line, const char * function) (assert.c:101)
- Eigen::internal::plain_array<double, 16, 0, 64>::plain_array(Eigen::internal::plain_array<double, 16, 0, 64> * const this) (\home\enahon\WORKSPACE\Dependencies\build\EP_INSTALL\include\eigen3\Eigen\src\Core\DenseStorage.h:147)
- Eigen::DenseStorage<double, 16, 4, 4, 0>::DenseStorage(Eigen::DenseStorage<double, 16, 4, 4, 0> * const this) (\home\enahon\WORKSPACE\Dependencies\build\EP_INSTALL\include\eigen3\Eigen\src\Core\DenseStorage.h:211)
- libMapEngine_lib_d.so!Eigen::PlainObjectBase<Eigen::Matrix<double, 4, 4, 0, 4, 4> >::PlainObjectBase<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 4, 4, 0, 4, 4> > >(Eigen::PlainObjectBase<Eigen::Matrix<double, 4, 4, 0, 4, 4> > * const this, const Eigen::DenseBase<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 4, 4, 0, 4, 4> > > & other) (\home\enahon\WORKSPACE\Dependencies\build\EP_INSTALL\include\eigen3\Eigen\src\Core\PlainObjectBase.h:590)
- libMapEngine_lib_d.so!Eigen::Matrix<double, 4, 4, 0, 4, 4>::Matrix<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 4, 4, 0, 4, 4> > >(Eigen::Matrix<double, 4, 4, 0, 4, 4> * const this, const Eigen::EigenBase<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 4, 4, 0, 4, 4> > > & other) (\home\enahon\WORKSPACE\Dependencies\build\EP_INSTALL\include\eigen3\Eigen\src\Core\Matrix.h:423)
- libgtsamDebug.so.4!gtsam::SO<4>::Hat(const gtsam::Vector6 & xi) (\home\enahon\WORKSPACE\Dependencies\vendor\GTSAM\gtsam-4.2a9\gtsam\geometry\SO4.cpp:58)
- libgtsamDebug.so.4!__static_initialization_and_destruction_0(int __initialize_p, int __priority) (\home\enahon\WORKSPACE\Dependencies\vendor\GTSAM\gtsam-4.2a9\gtsam\geometry\SO4.cpp:148)
- libgtsamDebug.so.4!_GLOBAL__sub_I_SO4.cpp(void)() (\home\enahon\WORKSPACE\Dependencies\vendor\GTSAM\gtsam-4.2a9\gtsam\geometry\SO4.cpp:234)
- ld-linux-x86-64.so.2!call_init(struct link_map * l, int argc, char ** argv, char ** env) (dl-init.c:70)
- ld-linux-x86-64.so.2!call_init(char ** env, char ** argv, int argc, struct link_map * l) (dl-init.c:33)
- ld-linux-x86-64.so.2!_dl_init(struct link_map * main_map, int argc, char ** argv, char ** env) (dl-init.c:117)
- ld-linux-x86-64.so.2!_dl_start_user (Unknown Source:0)
[Unknown/Just-In-Time compiled code] (Unknown Source:0)

**Comment by dellaert** [comment]
Dang, that's a hard one. Can you try with version 4.2a9?

**Comment by enahon** [comment]
Unfortunately I have already tried and it didn't really help. The stack trace above is actually from 4.2a9

**Comment by dellaert** [comment]
Unfortunately I'm no great help here, maybe @ProfFan can help.
One other thing to try is GTSAM_USE_SYSTEM_EIGEN=OFF

**Comment by thorstink** [comment]
I also experienced this if the march-flag was set differently for different libraries. In our production setup we compile all the libraries that use eigen from source with the same compilation settings: cpp17 to avoid eigen alignment issues and the same march-flag.If you compile your project with address sanitizer, and it points to some eigen function where you typically experience this double free, I think the above suggestion might help you.

**Comment by ProfFan** [comment]
I think we should convert this to discussion. @thorstink 's opinion is what I think is most likely to be the case. Remember that you have to make *ABSOLUTELY SURE* that your libraries are using the same settings for Eigen.

**Comment by dikokob** [comment]
@enahon, any update?

**Reply by enahon** [reply, marked as answer]
I have finally found a solution, following @thorstink's suggestion. PCL 1.13 uses -march=native by default, which causes conflicts if other dependencies also rely on Eigen but are being built with different flags.
To fix the issue, all dependencies need to be built with the same march setting. In my case I disabled it completely. For PCL, this is done by building from source with the following Cmake options:
        -DPCL_ENABLE_AVX=OFF
        -DPCL_ENABLE_SSE=OFF
        -DPCL_ENABLE_MARCHNATIVE=OFF

**Reply by GDTR12** [reply]
yes! it works well on my system
